src/vpn-api/worker.ts

```diff
@@ -71,91 +46,191 @@ const getInstanceSubnet = (instanceId: number) => {
+			timeout: 30000,
+			headers: { Authorization: `Bearer ${data.password}` },
+		})
+			.then(response => {
+				if (response.statusCode === 200) {
+					logger.debug(
+						`authentication passed for client_id=${clientId} uuid=${
+							data.username
+						}`,
+					);
+					return vpn
+						.exec(`client-auth-nt ${clientId} ${keyId}`)
+						.then(() =>
+							logger.info(
+								`authorised client_id=${clientId} uuid=${data.username}`,
+							),
+						);
+				} else {
+					logger.debug(
+						`authentication failed for client_id=${clientId} uuid=${
+							data.username
+						}`,
+					);
+					return vpn
+						.exec(
+							`client-deny ${clientId} ${keyId} "credentials rejected by api"`,
+						)
+						.then(() =>
+							logger.info(
+								`rejected client_id=${clientId} uuid=${data.username}`,
+							),
+						);
+				}
+			})
+			.catch(err => {
+				captureException(err, 'API Auth Error');
+			});
+	});
+
+	vpn.on('client:established', (clientId, data) => {
+		logger.info(
+			`connection established with client_id=${clientId} uuid=${data.username}`,
+		);
+		metrics.gauge('vpn_online_devices', vpn.nClients);
+		return clients.connected(data).then(logStateUpdate);
+	});
+
+	vpn.on('client:bytecount', (_clientId, data) => {
+		metrics.histogram('vpn_session_rx_bytes', data.bytes_received);
+		metrics.histogram('vpn_session_tx_bytes', data.bytes_sent);
 	});
 
-	const app = (Promise.promisifyAll(express()) as any) as AsyncApplication;
-	app.disable('x-powered-by');
-	app.get('/ping', (_req, res) => res.send('OK'));
-	app.use(morgan('combined'));
-	app.use(compression());
-	app.use(apiFactory());
-	app.use(Raven.errorHandler());
-
-	return app
-		.listenAsync(apiPort)
-		.tap(() =>
-			logger.info(
-				`open-balena-vpn worker-${instanceId} listening on port ${apiPort}`,
-			),
-		)
-		.tap(() => openvpn.start())
-		.tap(() =>
-			net.createConnection('/var/run/haproxy.sock', function(this: net.Socket) {
-				this.on('error', err => {
-					logger.error(`Error connecting to haproxy socket: ${err.message}`);
-					process.exit(1);
-				});
-				const preamble = `set server vpn-cluster/vpn${instanceId}`;
-				this.write(
-					`${preamble} addr 127.0.0.1 port ${vpnPort}\r\n${preamble} state ready\r\n`,
-					() => this.destroy(),
-				);
-			}),
-		)
-		.return(true);
+	vpn.on('client:disconnect', (clientId, data) => {
+		logger.info(
+			`session ended for client_id=${clientId} uuid=${data.username} (len=${
+				data.time_duration
+			} rx=${data.bytes_received} tx=${data.bytes_sent})`,
+		);
+		metrics.gauge('vpn_online_devices', vpn.nClients);
+		if (data.bytes_received != null && data.bytes_sent != null) {
+			metrics.histogram('vpn_session_rx_bytes', data.bytes_received);
+			metrics.histogram('vpn_session_tx_bytes', data.bytes_sent);
+		}
+		return clients.disconnected(data).then(logStateUpdate);
+	});
+
+	// setup metrics for prometheus
+	const kb = 2 ** 10; // 1024
```

Comment:
@Page- is going to ask for these to be explicit instead of powers of. :D

Reply:
Haha, the comment is there as an accepted compromise after a review from Page (in an earlier, ummerged, PR which this is superseding).

Reply:
Fair enough, I retract the comment. ;-)